File: Pouring1/Pouring.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Code;
using Flinq;

namespace Pouring1
{
    public partial class Pouring
    {
        private readonly int[] _capacities;
        private readonly State _initialState;
        private readonly Path _initialPath;
        private readonly IList<Move> _allPossibleMoves;

        public Pouring(params int[] capacities)
        {
            _capacities = capacities;
            _initialState = new State(_capacities.Map(_ => 0));
            _initialPath = new Path(this, _initialState, System.Linq.Enumerable.Empty<Move>());

            var glasses = System.Linq.Enumerable.Range(0, _capacities.Length).ToList();
            var moves1 = glasses.Map(g => new Empty(g));
            var moves2 = glasses.Map(g => new Fill(this, g));
            var moves3 = glasses.FlatMap(g1 => glasses.Map(g2 => (g1 != g2) ? new Pour(this, g1, g2) : null)).Where(x => x != null);

            _allPossibleMoves = moves1
                .Cast<Move>()
                .Concat(moves2)
                .Concat(moves3)
                .ToList();
        }

        public class State : List<int>
        {
            public State(IEnumerable<int> collection)
                : base(collection)
            {
            }

            public override string ToString()
            {
                return this.MkString("(", ",", ")");
            }
        }

        public abstract class Move
        {
            public abstract State Change(State state);
        }

        public class Path
        {
            private readonly Pouring _pouring;
            public State EndState { get; private set; }
            private readonly IEnumerable<Move> _history;

            public Path(Pouring pouring, State endState, IEnumerable<Move> history)
            {
                _pouring = pouring;
                EndState = endState;
                _history = history;
            }

            public Path Extend(Move move)
            {
                return new Path(_pouring, move.Change(EndState), new[] {move}.Concat(_history));
            }

            public override string ToString()
            {
                var movesAndStates = _history.FoldRight(
                    new List<Tuple<Move, State, string>>(),
                    (move, acc) =>
                        {
                            var previousState = (acc.IsEmpty()) ? _pouring._initialState : acc.First().Item2;
                            var nextState = move.Change(previousState);
                            var paddingPrefix = new string(' ', acc.Count);
                            var moveDescription = string.Format("{0}{1} => {2}", paddingPrefix, move, nextState);
                            return new[] {Tuple.Create(move, nextState, moveDescription)}.Concat(acc).ToList();
                        });
                return movesAndStates
                    .Map(x => x.Item3)
                    .Reverse()
                    .MkString(Environment.NewLine);
            }
        }

        private class StateEqualityComparer : IEqualityComparer<State>
        {
            public bool Equals(State x, State y)
            {
                return x.SequenceEqual(y);
            }

            public int GetHashCode(State x)
            {
                return x.GetHashCode();
            }
        }

        private static readonly IEqualityComparer<State> Comparer = new StateEqualityComparer();

        private Stream<IList<Path>> From(IList<Path> paths, IEnumerable<State> explored)
        {
            if (paths.IsEmpty()) return Stream<IList<Path>>.EmptyStream;
            var morePaths = paths
                .FlatMap(p => _allPossibleMoves.Map(p.Extend))
                .Where(p => !System.Linq.Enumerable.Contains(explored, p.EndState, Comparer)).ToList();
            return Stream<IList<Path>>.ConsStream(
                paths,
                () => From(
                    morePaths,
                    explored.Concat(morePaths.Map(p => p.EndState))));
        }

        public IEnumerable<Path> Solutions(int target)
        {
            var pathSets = From(new[] {_initialPath}, new[] {_initialState});
            return pathSets
                .ToEnumerable()
                .SelectMany(pathSet => pathSet.Where(path => path.EndState.Contains(target)));
        }
    }
}

File: Pouring.scala
package Coursera.ProgFun

class Pouring(capacities: Vector[Int]) {

  type State = Vector[Int]
  val initialState = capacities map (_ => 0)
  val initialPath = new Path(initialState, Nil)

  trait Move {
    def change(state: State): State
  }
  case class Empty(glass: Int) extends Move {
    def change(state: State): State = state updated (glass, 0)
  }
  case class Fill(glass: Int) extends Move {
    def change(state: State): State = state updated (glass, capacities(glass))
  }
  case class Pour(from: Int, to: Int) extends Move {
    def change(state: State): State = {
      val toCapacity = capacities(to)
      val fromState = state(from)
      val toState = state(to)
      val amount = fromState min (toCapacity - toState)
      state updated (from, fromState - amount) updated (to, toState + amount)
    }
  }

  val glasses = 0 until capacities.length
  val moves =
    (for (glass <- glasses) yield Empty(glass)) ++
    (for (glass <- glasses) yield Fill(glass)) ++
    (for (from <- glasses; to <- glasses; if from != to) yield Pour(from, to))

  class Path(val endState: State, history: List[Move]) {
    def extend(move: Move) = new Path(move change endState, move :: history)
    override def toString = {
      val movesAndStates = (history foldRight List[(Move, State, String)]()) ((move, acc) => {
        val previousState = acc match {
          case Nil => initialState
          case (_, ps, _) :: _ => ps
        }
        val nextState = move change previousState
        (move, nextState, s"$move => $nextState") :: acc
      })
      movesAndStates.map(x => x._3).reverse.mkString("[", ", ", "]")
    }
  }

  def from(paths: Set[Path], explored: Set[State]): Stream[Set[Path]] =
    if (paths.isEmpty) Stream.empty
    else {
      val morePaths = for {
        path <- paths
        next <- moves map path.extend
        if !(explored contains next.endState)
      } yield next
      paths #:: from(morePaths, explored ++ (morePaths map (_.endState)))
    }

  val pathSets = from(Set(initialPath), Set(initialState))

  def solutions(target: Int): Stream[Path] =
    for {
      pathSet <- pathSets
      path <- pathSet
      if path.endState contains target
    } yield path
}

object PouringApp {
  def main(args: Array[String]): Unit = {
    new Pouring(Vector(4, 9)).solutions(6).toList.foreach(println)
  }
}
